File: src/main/java/cope/nebula/client/feature/module/movement/NoRotate.java
package cope.nebula.client.feature.module.movement;

import cope.nebula.client.events.PacketEvent;
import cope.nebula.client.events.PacketEvent.Direction;
import cope.nebula.client.feature.module.Module;
import cope.nebula.client.feature.module.ModuleCategory;
import cope.nebula.client.value.Value;
import cope.nebula.util.world.entity.player.rotation.Rotation;
import net.minecraft.network.play.client.CPacketPlayer.PositionRotation;
import net.minecraft.network.play.server.SPacketPlayerPosLook;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

public class NoRotate extends Module {
    public NoRotate() {
        super("NoRotate", ModuleCategory.MOVEMENT, "Stops the server from rotating you");
    }

    public static final Value<Boolean> confirm = new Value<>("Confirm", true);

    @SubscribeEvent
    public void onPacket(PacketEvent event) {
        if (event.getPacket() instanceof SPacketPlayerPosLook && event.getDirection().equals(Direction.INCOMING)) {
            if (confirm.getValue()) {
                Rotation rotation = getNebula().getRotationManager().getFixedRotations();

                mc.player.connection.sendPacket(new PositionRotation(
                        mc.player.posX, mc.player.getEntityBoundingBox().minY, mc.player.posZ,
                        rotation.getYaw(), rotation.getPitch(), mc.player.onGround));
            }

            event.setCanceled(true);
        }
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/module/miscellaneous/NoRotate.kt
package wtf.nebula.client.feature.module.miscellaneous

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.network.play.client.CPacketConfirmTeleport
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraft.network.play.server.SPacketPlayerPosLook
import wtf.nebula.client.event.packet.PacketReceiveEvent
import wtf.nebula.client.event.packet.PacketSendEvent
import wtf.nebula.client.feature.module.Module
import wtf.nebula.client.feature.module.ModuleCategory

class NoRotate : Module(ModuleCategory.MISCELLANEOUS, "fuck ncp") {
    private var server = false
    private var confirm = false

    private var yaw = 0.0f
    private var pitch = 0.0f

    @EventListener
    private val packetSendListener = listener<PacketSendEvent> {
        if (it.packet is CPacketConfirmTeleport) {
            if (server) {
                confirm = true
            }
        }

        else if (it.packet is CPacketPlayer.PositionRotation) {
            if (server && confirm) {
                it.packet.yaw = yaw
                it.packet.pitch = pitch

                yaw = 0.0f
                pitch = 0.0f

                server = false
                confirm = false
            }
        }
    }

    @EventListener
    private val packetReceiveListener = listener<PacketReceiveEvent> {
        if (it.packet is SPacketPlayerPosLook) {
            val packet = it.packet

            yaw = packet.getYaw()
            pitch = packet.getPitch()

            server = true

            if (packet.flags.contains(SPacketPlayerPosLook.EnumFlags.Y_ROT)) {
                yaw += mc.player.rotationYaw
            }

            if (packet.flags.contains(SPacketPlayerPosLook.EnumFlags.X_ROT)) {
                pitch += mc.player.rotationPitch
            }

            packet.yaw = mc.player.rotationYaw
            packet.pitch = mc.player.rotationPitch
        }
    }
}